English - Basics/Containers (#1)
haskellfoundation/HaskellSchool: en/lessons/basics/containers.md

```haskell
ghci> head []
*** Exception: Prelude.head: empty list
ghci> tail []
*** Exception: Prelude.tail: empty list
```

We can use a common idiom in Haskell for covering partial functions in a safe
way, the `Maybe` type. This allows us to say that unhandled inputs return a
`Nothing`. Now the caller of this maybe-returning-function must handle the
`Nothing` case, but in return they are not faced with a nasty runtime exception.

```haskell
ghci> :i Maybe
data Maybe a = Nothing | Just a 	-- Defined in ‘GHC.Maybe’
...
```
__Note__: `:i` in ghci will give you some information about the type, the first
line is the implementation.

Now we can define a total head and tail function using pattern matching!

```haskell
ghci> :{
| safeHead :: [a] -> Maybe a
| safeHead [] = Nothing
| safeHead (x:xs) = Just x
|
| safeTail :: [a] -> Maybe [a]
| safeTail [] = Nothing
| safeTail (x:xs) = Just xs
| :}
ghci> safeHead ["Orange", "Banana", "Apple"]
Just "Orange"
ghci> safeHead []
Nothing
ghci> safeTail ["Orange", "Banana", "Apple"]
Just ["Banana","Apple"]
ghci> safeTail []
Nothing
```
__Note__: `:{` and `:}` allow you to write multiline definitions in ghci.

Hooray! No more exceptions.

> Another container that is the cousin of the list, and ensures that `head` and
> `tail` are safe is the nonempty list.
> 
> ```haskell
> ghci> :i NE.NonEmpty
> type NE.NonEmpty :: * -> *
> data NE.NonEmpty a = a NE.:| [a]
>   	-- Defined in ‘GHC.Base’

Reviewer: ```suggestion
Another way to ensures that `head` and `tail` are safe is the non-empty list:

```haskell
ghci> import Data.List.NonEmpty
ghci> :i NonEmpty
data NonEmpty a = a .:| [a]
```

```

From its definition we can see that `NonEmpty` requires the first element to be
present. This symbol `:|` is a constructor just like `:`, in fact the definition of
`NonEmpty` is identical to that of lists, except it omits the `[]` (Nil) case.

This handles the partiality problem the exact opposite way as the
`Maybe` solution. Instead of forcing the caller of the function to guard against
the ill-defined case when handling the result of the function, it forces them to
construct a valid input up front (when calling the function).

> ```haskell
> ghci> :t NE.head
> NE.head :: NE.NonEmpty a -> a
> ghci> NE.head (1 NE.:| [])
> 1
> ghci> NE.head []
> 
> <interactive>:11:9: error:
>     • Couldn't match expected type: NE.NonEmpty a
>                   with actual type: [a0]
>     ...

Reviewer: ```suggestion
```haskell
ghci> import qualified Data.List.NonEmpty as NE
ghci> :t NE.head
NE.head :: NonEmpty a -> a
ghci> head (1 :| [])
1
ghci> NE.head []

<interactive>:11:9: error:
    • Couldn't match expected type: NonEmpty a
                  with actual type: [a]
    ...
```

of two existing data types, the only thing we need is a lookup function, which
is provided in the `Data.List` module in base.

```haskell
ghci> import qualified Data.List as List
ghci> assoc = [("foo", True), ("bar", False)]
ghci> :t assoc
assoc :: [(String, Bool)]
ghci> :t List.lookup
List.lookup :: Eq a => a -> [(a, b)] -> Maybe b
ghci> List.lookup "foo" assoc
Just True
ghci> List.lookup "bar" assoc
Just False
ghci> List.lookup "baz" assoc
Nothing
```

We can see the pattern, again, where no result is encoded using the `Maybe`
type. This is because it is always *possible* that the key are looking up isn't
present in our assoc list.

It is also interesting to note the `Eq a` constraint on the "key" which allows
the lookup function to do an equality comparison on them.

While assoc lists are a nice introduction to key value containers, and they build
on the previous types we learned about, they are not particularly useful. A list
simply isn't a very good data structure for lookup, as it provides worst case
`𝛰(n)` asymptotics. Assoc lists are usually an intermediate data structure
which Haskell programmers will convert into a `Map`. Although this

> conversion is itself an `𝛰(n*log n)` operation `Map` provides an `O(log n)`

Reviewer: ```suggestion
conversion is itself an `𝛰(n*log n)` operation `Map` provides an `𝛰(log n)`
```

lookup, so the conversion cost is amortised very quickly.

## Sets

Sets are a very interesting container, the core concept of a set is membership.
It is common to build up sets, and then test to see if an element exists in that
set.

A set can be constructed exclusively by inserting elements into the empty set

> ```haskell
> ghci> :t Set.empty
> Set.empty :: Set.Set a
> ghci> Set.empty
> fromList []
> ghci> Set.insert 1 (Set.insert 2 (Set.insert 3 Set.empty))
> fromList [1,2,3]
> ```

Reviewer: ```suggestion
```haskell
ghci> import Data.Set (Set)
ghci> import qualified Data.Set as Set
ghci> :t Set.empty
Set.empty :: Set a
ghci> Set.empty
fromList []
ghci> Set.insert 1 (Set.insert 2 (Set.insert 3 Set.empty))
fromList [1,2,3]
```
```

Or by creating a set from a list

```haskell
ghci> Set.fromList [4,3,2,1]
fromList [1,2,3,4]
```

You might notice that the elements are sorted after turning them into a `Set`.
Internally the `Set` data type depends on its contents being orderable. The
concrete implementation of sets in Haskell are binary trees, which depend on an
ordering. We can see the ordering requirement from the constraints on the
`insert` and `fromList` (and other) functions.

```haskell

> ghci> :t Set.insert
> Set.insert :: Ord a => a -> Set.Set a -> Set.Set a
> ghci> :t Set.fromList
> Set.fromList :: Ord a => [a] -> Set.Set a
> ghci> :t Set.member
> Set.member :: Ord a => a -> Set.Set a -> Bool

Reviewer: ```suggestion
ghci> :t Set.insert
Set.insert :: Ord a => a -> Set a -> Set a
ghci> :t Set.fromList
Set.fromList :: Ord a => [a] -> Set a
ghci> :t Set.member
Set.member :: Ord a => a -> Set a -> Bool
```

for a hashmap. A hashmap simply hashes the key, et voilà, we have an ordering!
This does require that the key type is hashable though.

The module that exports the `HashMap` data type and functionality is called
`Data.HashMap.Strict`, it lives in the `unordered-containers` package. The api
is identical to `Map` aside from a `Hashable k` constraint instead of an `Ord k`
on the key.

## Vectors

Sometimes we want a list like container where we can have performant indexed
access. In fact arrays are often the primitive sequential data structure in
lots of popular programming langauges. Haskell has this kind of data structures
too, and the most popular implementation is the `Vector` type from the `vector`
library.

One of the coolest things about the `Vector` type is that it offers `𝛰(1)`
indexed access. The `vector` library offers both a safe and unsafe index
accessor operator.

Here is an example where we want to create a container that allows us to
retrieve ascii chars by their character code.

__Note__: You can fetch the `vector` library using `cabal repl --build-depends "vector"`
or `stack exec --package vector -- ghci`

```haskell

> ghci> import qualified Data.Vector as V

Reviewer: ```suggestion
ghci> import Data.Vector
ghci> import qualified Data.Vector as V
```

ghci> asciiChars = V.fromList ['\NUL'..'\DEL']
ghci> asciiChars ! 48
'0'
ghci> asciiChars ! 97
'a'
ghci> asciiChars ! 65
'A'
ghci> asciiChars !? 65
Just 'A'
ghci> asciiChars !? 128
Nothing
ghci> asciiChars !? 127
Just '\DEL'
```

Vectors also offer efficient slicing `𝛰(1)`, so they are really good for
creating sub-vectors. The slice function takes the index to start at, the
length of the slice, and the vector to slice, and returns a new vector. Slicing
is unsafe, if the starting index plus the length of the slice exceeds the vector
you will get a runtime error.

```haskell
ghci> :t V.slice
V.slice :: Int -> Int -> V.Vector a -> V.Vector a
ghci> lowerCase = V.slice 97 26 asciiChars
ghci> lowerCase
"abcdefghijklmnopqrstuvwxyz"
ghci> upperCase = V.slice 65 26 asciiChars
ghci> upperCase
"ABCDEFGHIJKLMNOPQRSTUVWXYZ"
ghci> nums = V.slice 48 10 asciiChars
ghci> nums
"0123456789"

>  ghci> V.slice 97 92 asciiChars -- Error case

Reviewer: I feel like you could say a bit more than "Error case". Like "Here, we attempt <forbidden operation>, and it fails"

"*** Exception: ./Data/Vector/Generic.hs:408 (slice): invalid slice (97,92,128)
CallStack (from HasCallStack):
  error, called at ./Data/Vector/Internal/Check.hs:87:5 in vector-0.12.3.0-8cc976946fcdbc43a65d82e2ca0ef40a7bb90b17e6cc65c288a8b694f5ac3127:Data.Vector.Internal.Check
```

## Overloaded Lists

You may have noticed that we frequently use the `fromList` function to
construct our containers. There is a language extension that allows us to
simply use the list syntax to construct these containers. The drawback of
overloading the list syntax is that we can get confusing type errors if we are
not explicit about our types.

```haskell
ghci> :set -XOverloadedLists -- This is the language extension

> ghci> [1,2,3] :: Set.Set Int
> fromList [1,2,3]
> ghci> [1,2,3] :: V.Vector Int
> [1,2,3]
> ghci> [('a', 1),('b', 2),('c', 3)] :: Map.Map Char Int
> fromList [('a',1),('b',2),('c',3)]

Reviewer: ```suggestion
ghci> [1,2,3] :: Set Int
fromList [1,2,3]
ghci> [1,2,3] :: Vector Int
[1,2,3]
ghci> [('a', 1),('b', 2),('c', 3)] :: Map Char Int
fromList [('a',1),('b',2),('c',3)]
```

```

However, if we aren't explicit about the type...

```haskell
ghci> ["1", "2", "3"]

<interactive>:11:1: error:
    • Illegal equational constraint GHC.Exts.Item l ~ [Char]
      (Use GADTs or TypeFamilies to permit this)
    • When checking the inferred type
        it :: forall {l}.
              (GHC.Exts.IsList l, GHC.Exts.Item l ~ [Char]) =>
              l
```

This error is a bit of a doozey, which is why this language extension is

> typically not enabled by default. It is still good to be aware of it!

Reviewer: Good thing that you show how it can fail.

---
version: 1.3.1
title: Containers
---

Tuples, Lists, Assoc Lists, Sets, Maps/Hashmaps, and Vectors.

{% include toc.html %}

## When Should I Use Which Container?

Here is a quick sentence about when to use each container. This is meant to be
a quick reference, and will probably make more sense after reading about each
container in more depth.

- Quickly group several pieces of data? Consider a [Tuple](#tuples)
- Fast prepend? Consider a [List](#lists)
- Unique and sorted elements? Consider a [Set](#sets)
- Lookup a key in a Key-Value mapping? Consider a [Map](#maps)
- Fast indexing? Consider a [Vector](#vectors)

## Using Container Modules

A very nice feature of the container ecosystem in Haskell is that all the
modules expose the same (or a very similar) api. This allows us to leverage our
intuition and feel familiar with a wide set of tools very quickly.

It also means that we are bound to run into namespace collisions. It is common
practice to import container modules using qualified imports. If you intend on
following along with your own `ghci` session, you should enter the qualified
imports below to make sure that all the modules are available in scope.

```haskell

> ghci> import qualified Data.Set as Set
> ghci> import qualified Data.Map as Map
> ghci> import qualified Data.List.NonEmpty as NE

Reviewer: ```suggestion
ghci> import Data.Set as Set
ghci> import Data.Map as Map
ghci> import Data.List.NonEmpty as NE
```

This works on the REPL and you have unqualified types :)

title: Collections
---

Tuples, Lists, Assoc Lists, Sets, Maps/Hashmaps, and Vectors.

{% include toc.html %}

## Tuples

Tuples are the first collection you are introduced to in Haskell. They are
simple, primitive, and have a terse, flexible, built in syntax. Tuples are
sometimes referred to as anonymous records, but instead of referencing fields
by name you reference them by position in the structure. Although tuples can
theoretically contain an arbitrary number of items, the reality is that the
Haskell Report only requires up to 15-tuples and GHC comes with support for
62-tuples. We will focus on the 2-tuple since Haskell has good built in support
for them. However, if you are interested this is what an 8-tuple looks like

```haskell
ghci> :t ('0', '1', '2', '3', '4', '5', '6', "8-tuple")
('0', '1', '2', '3', '4', '5', '6', "8-tuple")
  :: (Char, Char, Char, Char, Char, Char, Char, [Char])
```
__Note__: `:t` in the ghci repl shows the type of the value, it will print out a
message of the form `term :: Type`

### When to Use

Tuples are not very flexible because they have a fixed length. This also means
we have a lot of information about the structure of our data and therefore

> access is very perfomant `O(1)`. From a type safety perspective tuples are

Reviewer 2: ```suggestion
access is very performant `O(1)`. From a type safety perspective tuples are
```

useful for associating data; saying "I required both of these things". For
example if we wanted to represent a money transfer it wouldn't make sense to
have just the amount or just the destination, we would want to group them `(Int,
AccountId)`.

There is an interesting parallel between functions that require n arguments and
functions that require n-tuples. While "curried" functions are arguably more
useful due to partial application, these two modalities are equivalent.

```haskell
ghci> :t (&&)
(&&) :: Bool -> Bool -> Bool
ghci> :t uncurry (&&)
uncurry (&&) :: (Bool, Bool) -> Bool
ghci> :t curry (uncurry (&&))
curry (uncurry (&&)) :: Bool -> Bool -> Bool
```

### Constructing Tuples

> We can construct tuples using parnetheses with the elements separated by a

Reviewer 2: ```suggestion
We can construct tuples using parentheses with the elements separated by a
```

of two existing data types, the only thing we need is a lookup function, which
is provided in the `Data.List` module in base.

```haskell
ghci> assoc = [("foo", True), ("bar", False)]
ghci> :t assoc
assoc :: [(String, Bool)]
ghci> :t lookup
lookup :: Eq a => a -> [(a, b)] -> Maybe b
ghci> lookup "foo" assoc
Just True
ghci> lookup "bar" assoc
Just False
ghci> lookup "baz" assoc
Nothing
```

We can see the pattern, again, where undefined behaviour is encoded using the
`Maybe` type. This allows us to have complete functions with errors at compile
time rather than runtime.

It is also interesting to note the `Eq a` constraint on the "key" which allows
the lookup function to do an equality comparison on them.

While assoc lists are a nice introduction to key value collections that build
on the previous types we learned about, they are not particularly useful. A list
simply isn't a very good data structure for lookup, as it provides worst case
`O(n)` asymptotics. Assoc lists are usually an intermediate data structure
which Haskell programmers will usually convert into a `Map`. Although this
conversion is itself an `O(n*log n)` operation `Map` provides an `O(log n)`

> lookup, so the conversion cost is ammortized very quickly.

Reviewer 2: ```suggestion
lookup, so the conversion cost is amortised very quickly.
```
This uses the recommended British spelling.

## Sets

Sets are a very interesting container, the core concept of a set is membership.
Although you can lookup elements by index, this is generally not advised since
the interface makes no promises about what this index will be. Instead it is
common to build up sets, and then test to see if an element exists in that set.

A set can be constructed purely by inserting elements into the empty set

```haskell
ghci> import Data.Set
ghci> :t empty
empty :: Set a
ghci> empty
fromList []
ghci> insert 1 (insert 2 (insert 3 empty))
fromList [1,2,3]
```

Or by creating a set from a list

```haskell
ghci> fromList [4,3,2,1]
fromList [1,2,3,4]
```

You might notice that the elements are sorted after turning them into a `Set`.
Internally the `Set` data type depends on its contents being orderable. The

> concrete implementaiton of sets in haskell are binary trees, which depend on an

Reviewer 2: ```suggestion
concrete implementation of sets in Haskell are binary trees, which depend on an
```

ordering. We can see the ordering requirement from the constraints on the
`insert` and `fromList` (and other) functions.

```haskell
ghci> :t insert
insert :: Ord a => a -> Set a -> Set a
ghci> :t fromList
fromList :: Ord a => [a] -> Set a
ghci> :t member
member :: Ord a => a -> Set a -> Bool
```

Sets have a very useful property, they cannot contain duplicates. This means
that `insert` is idempotent.

```haskell
ghci> insert1 = insert 1
ghci> insert1 empty
fromList [1]
ghci> insert1 (insert1 empty)
fromList [1]
ghci> insert1 (insert1 (insert1 empty))
fromList [1]
```
__Note__: Idempotence is a property the some functions have. The property

> characterized by a function the yields the same result no matter how many times

Reviewer 2: ```suggestion
characterised by a function the yields the same result no matter how many times
```
This uses the recommended British spelling.

you call it (with the same argument). For idempotent function f `f a == f (f
a)`.

Alright, let's see an almost-practical use case for sets.

```haskell
ghci> :t member
member :: Ord a => a -> Set a -> Bool
ghci> evens = fromList [0,2..1000000]
ghci> member 7 evens
False
ghci> member 200012 evens
True
ghci> isEven n = member n evens
ghci> isEven 7
False
ghci> isEven 8
True
```

You might say "hmmm a `1000000` limit for even numbers seems incorrect",

> and you would be right! This highlights a property of sets in haskell, they

Reviewer 2: ```suggestion
and you would be right! This highlights a property of sets in Haskell, they
```

In regular arithmetic this relationship still holds.

```
a - (b - c) = (a + c) - b
4 - (2 - 1) = (4 + 1) - 2
3 = 3
```

But it doesn't work for our example.

```haskell
ghci> difference (union alphabet $ singleton 'y') vowels
fromList "bcdfghjklmnpqrstvwxz"
```

This is because `union alphabet $ singleton 'y'` is actually the same set as
`alphabet`; y is already a member of the alphabet set, so it is de-duplicated
when we try and union it in.

### Intersection

Intersection allows us to find elements that sets have in common.

```haskell
ghci> intersection vowels consonants
fromList "y"
```

### Subsets

Subsets allow us to detect if all the elements of a set are contained within
another set.

```haskell
ghci> isSubsetOf vowels alphabet
True
ghci> isSubsetOf consonants alphabet
True
ghci> isSubsetOf alphabet vowels
False
ghci> isSubsetOf consonants vowels
False
```

This function, like `difference`, has an order to the arguments. We are asking
if the first argument is a subset of the second. From the example above you can
see that `vowels` are a subset of the `alphabet` but not the other way around.

### Cartesian Products

> Cartesion products, sometimes referred to as cross multiplication, allow us to

Reviewer 2: ```suggestion
Cartesian products, sometimes referred to as cross multiplication, allow us to
```

A classic example of a cartesian product is a chess board, which is the cross
multiplication of the letters a - h and the numbers 1 - 8.

```haskell
ghci> ranks = fromList [1..8]
ghci> files = fromList ['a'..'h']
ghci> cartesianProduct files ranks
fromList
  [('a',1),('a',2),('a',3),('a',4),('a',5),('a',6),('a',7),('a',8)
  ,('b',1),('b',2),('b',3),('b',4),('b',5),('b',6),('b',7),('b',8)
  ,('c',1),('c',2),('c',3),('c',4),('c',5),('c',6),('c',7),('c',8)
  ,('d',1),('d',2),('d',3),('d',4),('d',5),('d',6),('d',7),('d',8)
  ,('e',1),('e',2),('e',3),('e',4),('e',5),('e',6),('e',7),('e',8)
  ,('f',1),('f',2),('f',3),('f',4),('f',5),('f',6),('f',7),('f',8)
  ,('g',1),('g',2),('g',3),('g',4),('g',5),('g',6),('g',7),('g',8)
  ,('h',1),('h',2),('h',3),('h',4),('h',5),('h',6),('h',7),('h',8)
  ]
```

You're can thank me later when you ace the n-queens question on your next
technical interview.

## Maps

In Haskell, maps are the "go-to" key-value store, sometimes referred to as a
dictionary.

A `Map` requires that its key has an ordering (`Ord k`), in the same way that
`Set` required an ordering on the value. This is because the internal

> implementation of the `Map` type in haskell is a size balanced binary tree. This

Reviewer 2: ```suggestion
implementation of the `Map` type in Haskell is a size balanced binary tree. This
```

in the snippet above.

Luckily `containers` exposes very uniform api's. So we get all of our set
operations on maps, and they are just as performant! Lists also have the set
operations but they are much slower due to the different performance
characteristics of linked lists and binary trees.

## Adding Values to a Map

The blessed function for adding to a map is called `insert`. If a duplicate key
is added to a map, it will replace the former value:

```haskell
ghci> empty
fromList []
ghci> oneItem = insert "first" 1 empty
ghci> oneItem
fromList [("first",1)]
ghci> insert "first" 2 oneItem
fromList [("first",2)]
```

A useful function to be aware of is `adjust`. This lets us update a value at a
specified key, only if it exists, if not the old map is returned.

```haskell
ghci> adjust (+2) "first" oneItem
fromList [("first",3)]
ghci> :t adjust
adjust :: Ord k => (a -> a) -> k -> Map k a -> Map k a
ghci> adjust (+2) "second" oneItem
fromList [("first",1)]
```

The observant reader will notice that `adjust` doesn't actually update the map.
The second invocation of adjust returns the original `oneItem` map, this makes

> sense when you consider that all data in haskell is immutable!

Reviewer 2: ```suggestion
sense when you consider that all data in Haskell is immutable!
```

           or ‘Prelude.lookup’,
              imported from ‘Prelude’ (and originally defined in ‘GHC.List’)
ghci> empty

<interactive>:5:1: error:
    Ambiguous occurrence ‘empty’
    It could refer to
       either ‘Data.Map.empty’,
              imported from ‘Data.Map’
              (and originally defined in ‘Data.Map.Internal’)
           or ‘Data.Set.empty’,
              imported from ‘Data.Set’
              (and originally defined in ‘Data.Set.Internal’)
```

## Tuples

Tuples are the first container you are introduced to in Haskell. They are
simple, primitive, and have a terse built-in syntax. The fields of a tuple are
referenced by position in the structure. Although tuples can theoretically
contain an arbitrary number of items, the reality is that the Haskell Report
only requires up to 15-tuples and GHC comes with support for 62-tuples. We will
focus on the 2-tuple since Haskell has good built in support for them. Just so
you can see, here is an example of a tuple with 8 fields.

```haskell
ghci> :t ('0', '1', '2', '3', '4', '5', '6', "8-tuple")
('0', '1', '2', '3', '4', '5', '6', "8-tuple")
  :: (Char, Char, Char, Char, Char, Char, Char, [Char])
```

> __Note__: `:t` in the ghci repl shows the type of the value, it will print out a

Reviewer 2: There is a comma splice here.

tuple is a distinct data type. That means there is no general function to add an
additional element to any tuple. Rather these kinds of functions need to be
defined ad hoc.

Here is an example of increasing the tuple's length.

```haskell
ghci> twoTupleToThreeTuple c (a, b) = (a, b, c)
ghci> twoTupleToThreeTuple () myTuple
(1, "world", ())
```

Trying to call our function on a tuple of the incorrect length results in a type
error telling us just as much; the function expected a 2-tuple but we gave it a
3-tuple.

```haskell
ghci> twoTupleToThreeTuple True (1, 2, 3)

<interactive>:19:27: error:
    • Couldn't match expected type: (a, b)
                  with actual type: (a0, b0, c0)
                  ...
```

## Lists

From a usability perspective, Lists solve the problem of extensibility that
tuples face, but they can only contain one type (in other words they are
homogeneous). Lists also have a special built in syntax.

```haskell
ghci> [1,2,3,4]
[1,2,3,4]
ghci> ["hello", True]

<interactive>:22:11: error:
    • Couldn't match type ‘Bool’ with ‘[Char]’
      Expected: String
        Actual: Bool
    ...
```

> 

Reviewer 3: I’m still confused  as to why would you put it here. There will be a chapter on user-defined datatypes, but I don’t think you need this intro here. And I don’t think it helps to understand what follows in any way.

Author: In another comment someone asked to see the error of having multiple types in a list. I think we are getting to a point where different opinions are at play. Would be interested what other people think about this, and I am happy to remove if that's the consensus.

Reviewer 2: I feel that, for beginners, we should try to give a terse overview of the language that gives them the impression that they could use the language themselves. Showing too many type errors may be off-putting as it slows down the pace of progress they feel they make and their sense of understanding of the language.

Reviewer: I agree with @TrueBoxGuy on this one, the pedagogical mechanics of "Do this in order to have that" (what we do) and "Here is how it can go wrong" are quite different (and the latter would be more suited for a bigger format, like a dedicated blog post)

Reviewer 3: Friends, I'm sorry: my comment was meant for the next section where a manual definition of list is given. I don't mind the examples of errors.

Author: Ahhh, I was trying to demonstrate how the list data type works under the hood (since the built in syntax kind of hides that). I thought lists were a good moment to introduce inductive / recursive types. If you think that is valuable, but wasn't clear, I can try and make it more explicit. If you think that this topic is out of place and distracting, I am happy to move it.

Reviewer 3: Yeah, i was arguing for the latter, that is: not having it in this document at all because user-defined types is a huge topic on itself. Any online tutorial on basic data structures I can remember  (tuples/lists first and foremost, and not only in Haskell, other languages too) shoes how to use lists but not how to define them. Understanding of data-definitions shouldn't be required to understand how to use lists in Haskell! Data-definitions should be the next step instead.

Reviewer 3: /cc @Kleidukos in case you missed this discussion about list definition.

Reviewer: We better take this discussion on a more appropriate medium like the Slack

Haskell implements lists as linked lists. The cons cells (the operator `:` is
called cons, short for constructor) act as the links. This dictates which
operations can be done quickly and which can be slow:

Prepending a value to a list is easy and fast - all we have to do is create a
new cons cell with the element we want to prepend and point it to the existing
list.

```haskell
prepend value list = value : list
```

On the other hand, since the list data type (as shown above) can be either
empty (`[]` or `Nil`) or a cons cell that will point to the rest of the list,
it does not contain information about the length of the list, or a reference to
the end of the list.

Because of that, in order to retrieve the length of a list we must walk
each cons cell and count until we reach the end of the list. To find the
value at a specific index we need to traverse the list until we reach it.

Similarly, in order to append a list to an existing list, we need to go to the
end of the existing list, and add a cons cell that points to the new list:

```haskell
append originalList newList =
    case originalList of
        [] -> newList
        x : xs -> x : append xs newList
```

> 

Reviewer 3: There should be a mention of `++` in particular, I think: its usage really is a code smell for anything more complex than hello ++ world. In particular, I’d mention as an example that having ++ in a loop brings you quadratic performance.

ordering. We can see the ordering requirement from the constraints on the
`insert` and `fromList` (and other) functions.

```haskell
ghci> :t Set.insert
Set.insert :: Ord a => a -> Set.Set a -> Set.Set a
ghci> :t Set.fromList
Set.fromList :: Ord a => [a] -> Set.Set a
ghci> :t Set.member
Set.member :: Ord a => a -> Set.Set a -> Bool
```

Sets have a very useful property, they cannot contain duplicates.

```haskell
ghci> insert1 = Set.insert 1
ghci> insert1 Set.empty
fromList [1]
ghci> insert1 (insert1 Set.empty)
fromList [1]
ghci> insert1 (insert1 (insert1 Set.empty))
fromList [1]
```

Alright, let's see an almost-practical use case for sets.

```haskell
ghci> evens = Set.fromList [0,2..1000000]
ghci> Set.member 7 evens
False
ghci> Set.member 200012 evens
True
ghci> isEven n = Set.member n evens
ghci> isEven 7
False
ghci> isEven 8
True
```

You might say "hmmm a `1000000` limit for even numbers seems incorrect", and you
would be right! This highlights a property of sets in haskell, they are finite
due to the strictness in the underlying data structure. This contrasts lists,
which are lazy and therefore potentially infinite.

> ### Difference

Reviewer 3: This series of set-related operations take incredibly large space for such a niche instrument: I rarely find myself taking set differences, for example. I think 20% of interface (really: insert and lookup) cover 80% of users.

More concretely, I’d suggest to pack the set-theoretic operations (diff, union, intersect and subset; maybe Cartesian product) into one section with a single block of examples. This is a drastic change, so maybe if others are really happy with the current version, I don’t want to insist. But I wholeheartedly believe this topic is currently vastly overweighted with little benefit.

Author: Would be interested to hear what others think

Reviewer 2: Shortening the page (by grouping operations together and using more terse language) would likely make it more digestible for readers and increase the speed of the progress they make.

Reviewer: As said in my other review, I would be absolutely delighted if the Set part could be a fully-fledged blog post on Haskell School. The point of these lessons is to give a touch and feel of the language, sometimes at the price of exhaustiveness.

Author: It seems there is consensus!

`Set` required an ordering on the value. This is because the internal
implementation of the `Map` type in Haskell is a size balanced binary tree. This
is actually the same data structure as is used for `Set`.

The `Map` data type and the functions for interacting with it are exported from
`Data.Map` which is a module in the `containers` package. Don't worry about
dependencies though, containers is a core library and ships with ghci.

Like sets, a map can be constructed by inserting key value pairs into an empty map

```haskell
ghci> :t Map.empty
Map.empty :: Map.Map k a
ghci> Map.empty
fromList []
ghci> Map.insert "a" 'a' (Map.insert "b" 'b' (Map.insert "c" 'c' Map.empty))
fromList [("a",'a'),("b",'b'),("c",'c')]
```

Or by creating it from a list

```haskell
ghci> Map.fromList [(4, '4'), (3, '3'), (2, '2'), (1,'1')]
fromList [(1,'1'),(2,'2'),(3,'3'),(4,'4')]
```

## Updating Values to a Map

A useful function to be aware of is `adjust`. This lets us update a value at a
specified key, only if it exists, if not the old map is returned.

```haskell
ghci> Map.adjust (+2) "first" oneItem
fromList [("first",3)]

> ghci> :t Mapadjust

Reviewer 2: ```suggestion
ghci> :t Map.adjust
```

Map.adjust :: Ord k => (a -> a) -> k -> Map.Map k a -> Map.Map k a
ghci> Map.adjust (+2) "second" oneItem
fromList [("first",1)]
```

The observant reader will notice that `adjust` doesn't actually update the map.
The second invocation of adjust returns the original `oneItem` map, this makes
sense when you consider that all data in Haskell is immutable!

## When to use Maps

Maps are really great for in memory persistence of state that will need to be
retrieved by a key of some arbitrary type. This is is because maps have great
lookup asymptotics (`𝛰(log n)`) due to the ordering on the key values. The
example that immediately springs to mind is session storage on the server in a
web application. The session state can be indexed by an id that is stored in
the cookie, and the session state can be retrieved from an in memory `Map` on
every request. This solution doesn't scale infinitely, but you would be
surprised how well it works!

> 

Reviewer 3: I’m not sure HashMaps are even worth mentioning in this supposedly introductory text. [There’s a recent post](https://oleg.fi/gists/posts/2021-05-19-dont-default-to-hashmap.html) by @phadej arguing that HashMaps are overrated...

Author: hmmmm, I would be interested in @Kleidukos thoughts here. Its a really small section and doesn't add that much, so I could see how it would be more confusing than beneficial.

Reviewer: HashMap is still a pretty big part of this ecosystem, despite its warts. Let us keep this section for now.
